### PR TITLE
[#6896] Requested replica always wins the vote (main)

### PR DESCRIPTION
**Pull request by alanking**
Addresses #6896

Core tests are now running

### CODE REVIEW

**Review by alanking — COMMENTED** [review]
`test_ichksum` and `irods_logical_locking` (unit test) were failing due to unexpected error messages. As it turns out, the solution offered by isocpp.org does not seem to work for our compiler(s)/platform(s) ("mileage may vary"). I tried this out on Compiler Explorer, and things were working as expected, but then I saw some tests failing because votes were not coming out as 0 anymore in instances where they should have.

I'm going to remove this in favor of another function - `vote_is_zero` - which does the safe comparison with `vote::zero` like we do in some (but not all) places now. Should we choose to implement said comparison with something like this function in the future, now it is hidden behind the one interface.

Confirmed that failing tests are now passing, but doing another full suite run now to be sure.

File: server/core/include/irods/voting.hpp
Comment (alanking) [inline server/core/include/irods/voting.hpp:28]: Remove this line

Comment (alanking) [reply server/core/include/irods/voting.hpp]: Done

**Comment by alanking** [comment]
Unit and core tests passed

**Comment by trel** [comment]
Nice - happy to see it squashed - this is very nice work.

**Comment by alanking** [comment]
Went ahead and did the squash. Sorry if that messed with any reviews!

File: server/api/src/rsDataObjOpen.cpp
Comment (korydraughn) [inline server/api/src/rsDataObjOpen.cpp:771]: Consider reformatting or refactoring to use if-statements.

Chained ternary operators can hurt readability.

File: server/core/src/voting.cpp
Comment (korydraughn) [inline server/core/src/voting.cpp:255]: So this is how we've always checked for zero, it's just that you've wrapped the logic so that it is in one location, correct?

File: scripts/irods/test/test_targeting_specific_replica_number.py
Comment (korydraughn) [inline scripts/irods/test/test_targeting_specific_replica_number.py:66]: run_icommand?

File: scripts/irods/test/test_targeting_specific_replica_number.py
Comment (korydraughn) [inline scripts/irods/test/test_targeting_specific_replica_number.py:15]: Are we favoring lowercase instead of uppercase for class names now?
Is there some motivating factor behind this?

File: scripts/irods/test/test_targeting_specific_replica_number.py
Comment (korydraughn) [inline scripts/irods/test/test_targeting_specific_replica_number.py:126]: What is `assert_command`? Is that part of the unittest framework or do we provide that?

File: scripts/irods/test/test_targeting_specific_replica_number.py
Comment (korydraughn) [inline scripts/irods/test/test_targeting_specific_replica_number.py:306]: mtime?

File: plugins/resources/src/passthru.cpp
Comment (alanking) [reply plugins/resources/src/passthru.cpp:78]: It returns an `irods::error` object. I'll add a check here for it.

File: server/core/src/voting.cpp
Comment (alanking) [reply server/core/src/voting.cpp:255]: Yes and no.

This is the way we checked for zero in several places in `irods_resource_redirect.cpp`. In other places, the vote value was compared directly against `0.0f` (or, more recently, `vote::zero`). You can see the specific instances elsewhere in the diff.

This function provides a thin abstraction so that the implementation can be swapped out later if we so choose. I think comparing against zero should be consistent, in any case.

File: scripts/irods/test/test_targeting_specific_replica_number.py
Comment (alanking) [reply scripts/irods/test/test_targeting_specific_replica_number.py:15]: As near as I can tell, the only reason we use uppercase letters in these classes is because you can't name the class the same thing as the file in the `unittest` framework.

I personally hate the mixed-case naming convention, which is why I opted for this scheme. I have done this several times with already-merged test classes, so I figured it was okay.

I am willing to change this to whatever everybody finds most appropriate.

File: scripts/irods/test/test_targeting_specific_replica_number.py
Comment (alanking) [reply scripts/irods/test/test_targeting_specific_replica_number.py:126]: We provide it: https://github.com/irods/irods/blob/ee3a689971435d81e12aef6500d27ff33a9c2a01/scripts/irods/test/command.py#L76-L77

File: scripts/irods/test/test_targeting_specific_replica_number.py
Comment (korydraughn) [reply scripts/irods/test/test_targeting_specific_replica_number.py:15]: It is fine. I only brought it up since I have seen different schemes used in the codebase. Likely an indicator that we need more convention, but certainly not an urgent matter right now.

Please carry on.

File: server/core/src/voting.cpp
Comment (alanking) [reply server/core/src/voting.cpp:255]: All of these comparisons have already been replaced by calls to `vote_is_zero` in this PR. I can put them back and make a new issue to address this at a later time if that seems better. Or are you referring to all floating point comparisons in the codebase?

File: scripts/irods/test/test_targeting_specific_replica_number.py
Comment (alanking) [reply scripts/irods/test/test_targeting_specific_replica_number.py:126]: I chose `assert_command` because `diff` is not an iCommand, so `self.<some_user>.assert_icommand` seemed unfitting (although I have no reason to believe it wouldn't do the same thing). There is no member of `IrodsSession` which serves an equivalent role, as far as I can tell.

File: server/core/src/voting.cpp
Comment (korydraughn) [reply server/core/src/voting.cpp:255]: No need to revert anything.

I was under the impression that there were other places outside of your PR that dealt with voting, but compared floats incorrectly. It sounds like you've already addressed all known locations though.

Comment (alanking) [reply server/core/src/voting.cpp:255]: Oh, I understand now. Yeah, I didn't really look outside of `irods_resource_redirect.cpp`. Looks like some of the resource plugins are making direct comparisons, so I made an issue to address those places: https://github.com/irods/irods/issues/6932

Comment (korydraughn) [reply server/core/src/voting.cpp:255]: Cool cool

File: plugins/resources/src/passthru.cpp
Comment (alanking) [reply plugins/resources/src/passthru.cpp]: Well, I went the other direction and am now using a throwing function. Please let me know if returning an `irods::error` seems more appropriate.

File: server/api/src/rsDataObjOpen.cpp
Comment (alanking) [reply server/api/src/rsDataObjOpen.cpp]: Done

File: server/api/src/rsDataObjOpen.cpp
Comment (alanking) [reply server/api/src/rsDataObjOpen.cpp]: Done

File: scripts/irods/test/test_targeting_specific_replica_number.py
Comment (alanking) [reply scripts/irods/test/test_targeting_specific_replica_number.py]: Done

File: scripts/irods/test/test_targeting_specific_replica_number.py
Comment (alanking) [reply scripts/irods/test/test_targeting_specific_replica_number.py]: Done

File: plugins/resources/src/passthru.cpp
Comment (alanking) [inline plugins/resources/src/passthru.cpp:820]: I would print the name of the resource here, but getting the name of the resource may be the very thing which causes the exception!

File: plugins/resources/src/passthru.cpp
Comment (korydraughn) [reply plugins/resources/src/passthru.cpp]: Will take a look.

File: plugins/resources/src/passthru.cpp
Comment (korydraughn) [reply plugins/resources/src/passthru.cpp:820]: Perhaps you can capture it earlier in the function?

File: plugins/resources/src/passthru.cpp
Comment (korydraughn) [reply plugins/resources/src/passthru.cpp]: Looks good to me.

**Review by korydraughn — COMMENTED** [review]
Changes look good. Anything else left to do for this?

File: plugins/resources/src/passthru.cpp
Comment (alanking) [reply plugins/resources/src/passthru.cpp:820]: Hmm... that's true. I'll play around with that.

**Comment by alanking** [comment]
Other than the pending comment, I need to run through tests one more time, I think.

**Comment by alanking** [comment]
Well, ran into some issues regarding the "prefer archive" policy configuration for the compound resource, resulting in a couple of test failures. I think we arrived at a couple of good options for solving that, and I have confirmed that the tests which were failing are now passing with those changes.

However, when I tried running the full test suite, I observed some other very unrelated tests failing (e.g. `test_imeta_help`) and, after observing failures in the tests for known-working commits, have concluded that my workstation is having issues. I'll follow up once I get the tests to pass.

**Comment by alanking** [comment]
Tests are now passing. I rebuilt the base images and things healed up. Latest commit shows the vote-settting logic.

File: plugins/resources/src/compound.cpp
Comment (trel) [inline plugins/resources/src/compound.cpp:1637]: "replica in the cache replica" -> "replica in the cache"

File: plugins/resources/src/compound.cpp
Comment (korydraughn) [inline plugins/resources/src/compound.cpp:1641]: Is `r` a pointer?

If it is, consider appending an asterisk to `auto` so that it is clear to the reader what `r` is.

Comment (korydraughn) [reply plugins/resources/src/compound.cpp:1641]: No it's fine. I figured it was either a pointer or something that offered pointer-like semantics.

Carry on.

File: plugins/resources/src/compound.cpp
Comment (alanking) [reply plugins/resources/src/compound.cpp]: Done

File: plugins/resources/src/compound.cpp
Comment (alanking) [reply plugins/resources/src/compound.cpp:1641]: Okey dokey. Resolving!

**Review by korydraughn — COMMENTED** [review]
If the tests are passing, please squash without the pound.

**Comment by alanking** [comment]
Squashed. This includes a rebase, so I will run tests again soon.

Plus, I promised @JustinKyleJames I would hold off on merging this until #6933 is merged since there may be conflicts :)

**Comment by alanking** [comment]
Rebased, running tests again

**Comment by alanking** [comment]
Unit tests and core tests passed

File: scripts/irods/lib.py
Comment (korydraughn) [inline scripts/irods/lib.py:771]: This could return a tuple containing the resc id and name. If not, then consider adding `_name` to the function name.

Feel free to ignore this.

Comment (alanking) [reply scripts/irods/lib.py:771]: I think I'd rather split it in two, in that case: `get_resource_parent_name` and `get_resource_parent_id`. Hold please.

Comment (alanking) [reply scripts/irods/lib.py:771]: Well, before I make the change, just know that this has already been merged in 4-2-stable as-is. Are we okay with the split?

Comment (korydraughn) [reply scripts/irods/lib.py:771]: Let's leave it as is.

Consider creating an issue to revisit this in main and 4-3-stable.

Comment (alanking) [reply scripts/irods/lib.py:771]: Yep, makes sense. Created https://github.com/irods/irods/issues/7029.

**Review by korydraughn — APPROVED** [review]
If you're happy with it, please pound and merge when ready.

**Pull request merged**
